Allow Scaffolder to accept props for custom `TemplateListPage` and `TemplateWizardPage`

minkimcello opened the pull request:
## Motivation 

Replaces #16873.

✅ ~~Should merge #16949 first.~~

## Approach

_✨🚨 Moved `TemplateGroupFilter` and `TemplateGroups` to the `scaffolder-react` package_

The alpha scaffolder plugin will now accept entire pages as props:

```diff
// ./packages/app/src/App.tsx

+ import { MyListPage } from './MyListPage';
+ import { MyWizardPage } from './MyWizardPage';

<Route
  path="/create"
  element={
    <NextScaffolderPage
      groups={[
        { ... },
      ]}
      components={{
+        TemplateListPageComponent: MyListPage,
+        TemplateWizardPageComponent: MyWizardPage,
      }}
    />
  }
>
```

In the screenshots below, I copied both components over to the example app and just updated the header title:

![Screenshot 2023-03-16 at 1 33 11 PM](https://user-images.githubusercontent.com/29791650/225704659-b0d179ef-e17c-44f4-a392-6b125c70ad9f.png)

- Some of the modules were not being exported that would allow a user to copy/paste the default list/wizard pages into their own app and replace the local import sources to `@backstage/plugin-scaffolder/alpha` so I went through each import to make sure they are exported
- In order to create your own `TemplateListPage` and `TemplateWizardPage`, users will have to add `@backstage/scaffolder-common` and `@backstage/types` to their app

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] ~~Added or updated documentation~~ _⚠️ The changes proposed are for the alpha version of the scaffolder_
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

jhaals commented:
Leaving this one in your capable hands @benjdlambert

benjdlambert commented on plugins/scaffolder/alpha-api-report.md:
```diff
@@ -16,6 +20,14 @@ import { ScaffolderTaskOutput } from '@backstage/plugin-scaffolder-react';
 import { TemplateEntityV1beta3 } from '@backstage/plugin-scaffolder-common';
 
+// @alpha
+export const CategoryPicker: () => JSX.Element | null;
```
Thinking that these components should be moved to `-react` instead and be exported from there. Doesn't make sense for them to live in the plugin anymore.

benjdlambert replied:
Also rename?

minkimcello replied:
How's `TemplateCategoryPicker`?

Created a separate PR: #16949

benjdlambert commented on plugins/scaffolder/alpha-api-report.md:
```diff
@@ -16,6 +20,14 @@ import { ScaffolderTaskOutput } from '@backstage/plugin-scaffolder-react';
+export const CategoryPicker: () => JSX.Element | null;
+
+// @alpha (undocumented)
+export function ContextMenu(
```
`@backstage/plugin-scaffolder-react`

benjdlambert replied:
Probably look for a better name here too for where it's used etc.

minkimcello replied:
[`ContextMenu` imports three route refs](https://github.com/backstage/backstage/blob/310b781eac12b52a031442b77e12a57ed5bc39b0/plugins/scaffolder/src/next/TemplateListPage/ContextMenu.tsx#L32-L36) from within the same package.

Should I convert the three boolean args into route ref args?

benjdlambert replied:
Yep, either that, or providing `onClick` handlers for each of them.

benjdlambert replied:
And I think that might be outdated now as those routeRefs don't exist anymore, they're just the same `routeRefs` as of the release tomorrow :)

benjdlambert commented on plugins/scaffolder/alpha-api-report.md:
```diff
@@ -60,11 +74,64 @@ export const nextSelectedTemplateRouteRef: SubRouteRef<
 >;
 
+// @public (undocumented)
+export const registerComponentRouteRef: ExternalRouteRef<undefined, true>;
```
Don't think that we should export this. We can just leave the authors to implement an explicit `routeRef` if they need one, or if they don't ignore and don't include the component. This was only done to make it configurable to show a button or not depending on if they bound the route ref in the plugin.

benjdlambert commented on plugins/scaffolder/alpha-api-report.md:
```diff
@@ -60,11 +74,64 @@ export const nextSelectedTemplateRouteRef: SubRouteRef<
+export const registerComponentRouteRef: ExternalRouteRef<undefined, true>;
+
+// @alpha
+export const RegisterExistingButton: (
```
I'm also in two minds to export this too - feels pretty small and maybe not worth it?

benjdlambert reviewed (changes requested):
Hey :wave:

Thanks for this - some comments for a first pass. Basically just moving some bits round and renaming the exports, also some questions about if it's work exporting some things.

benjdlambert commented:
@minkimcello I amended this PR a little bit on top of the previously merged refactor to move out some components. Realised that we should export the `TemplateGroups` component too, rather than it coming from the plugin package itself. So had to move that around and fix the props for that too. Thinking that this is looking pretty good now, and would like to get it in this release today if possible.

minkimcello commented:
> @minkimcello I amended this PR a little bit on top of the previously merged refactor to move out some components. Realised that we should export the TemplateGroups component too, rather than it coming from the plugin package itself. So had to move that around and fix the props for that too. Thinking that this is looking pretty good now, and would like to get it in this release today if possible.

Updated the PR and rebased. Everything looks good to me too.

minkimcello commented:
> Thank you for contributing to Backstage! The changes in this pull request will be part of the `1.13.0` release, scheduled for Tue, 18 Apr 2023.

😩

benjdlambert commented:
Relax @minkimcello it's going to be in the `-next` release today 😝

benjdlambert commented:
That's the next `stable` release.
